a3p-integration/proposals/p:upgrade-19/psm.test.js

```diff
@@ -0,0 +1,81 @@
+/* eslint-env node */
+/**
+ * @file The goal of this file is to make sure all the PSM contacts are upgrade. Each set of vats are different
+ * per chain, but for mainnet, we expect this to be: V37-V42, V73, V76
+ *
+ * The test scenario is as follows;
+ * 1. Simulate trade of IST and USDC
+ * 2. Upgrade all PSMs
+ * 3. Verify metrics are the same after the upgrade
+ * 4. Verify trading is still possible after the upgrade
+ */
+
+import '@endo/init';
+import test from 'ava';
+import { evalBundles } from '@agoric/synthetic-chain';
+import { makeVstorageKit } from '@agoric/client-utils';
+
+const UPGRADE_PSM_DIR = 'upgradePSM';
+const SWAP_ANCHOR = 'swapAnchorForMintedSeat';
+
+test.before(async t => {
+  const vstorageKit = await makeVstorageKit(
+    { fetch },
+    { rpcAddrs: ['http://localhost:26657'], chainName: 'agoriclocal' },
+  );
+
+  t.context = {
+    vstorageKit,
+  };
+});
+
+test.serial('simulate trade of IST and USDC', async t => {
+  // @ts-expect-error casting
+  const { vstorageKit } = t.context;
+
+  await evalBundles(SWAP_ANCHOR);
+
+  const metrics = await vstorageKit.readLatestHead(
+    'published.psm.IST.USDC.metrics',
+  );
+
+  t.is(metrics.anchorPoolBalance.value, 500000n);
+  t.is(metrics.feePoolBalance.value, 0n);
+  t.is(metrics.mintedPoolBalance.value, 500000n);
+  t.is(metrics.totalAnchorProvided.value, 0n);
+  t.is(metrics.totalMintedProvided.value, 500000n);
+});
+
+test.serial('upgrade PSMs', async t => {
+  // @ts-expect-error casting
+  const { vstorageKit } = t.context;
+
+  await evalBundles(UPGRADE_PSM_DIR);
+
+  const metrics = await vstorageKit.readLatestHead(
+    'published.psm.IST.USDC.metrics',
+  );
+
+  t.is(metrics.anchorPoolBalance.value, 500000n);
+  t.is(metrics.feePoolBalance.value, 0n);
+  t.is(metrics.mintedPoolBalance.value, 500000n);
+  t.is(metrics.totalAnchorProvided.value, 0n);
+  t.is(metrics.totalMintedProvided.value, 500000n);
```

Comment:
I find it hard to compare these values visually with the previous and following values. What I've done in the past is use a variable to hold the mutating value, so the comparison always looks the same, and the changes appear explicitly near the place that changes the value.

In [vaultFactoryTest](https://github.com/Agoric/agoric-sdk/blob/e06691f3bb13d315a4b557b4711c33cabc94979c/packages/inter-protocol/test/vaultFactory/vaultFactory.test.js#L1780), `totalCollateral` holds the evolving value, and it can be inserted in many different assertions 
```
  await m.assertChange({
    totalCollateral: { value: totalCollateral },
  });
```
When it changes, there's an explicit assignment. 

With 5 values changing in sync here, I'd define a record 
```
  const poolMetrics = {
    anchorBalance: 500_000n,
    feeBalance: 0n
    mintedBalance: 500_000n,
    anchorProvided: 0n,
    mintedProvided: 500_000n,
};
```
and a comparer `compareLevels(metrics, poolMetrics);`. Then we could see that there's only one place where we change the values of `poolMetrics`
```
  poolMetrics.anchorBalance = poolMetrics.anchorBalance + 500_000n;
  poolMetrics.mintedBalance = poolMetrics.mintedBalance + 500_000n;
  poolMetrics.mintedProvided = poolMetrics.mintedProvided + 500_000n;
```
and otherwise, the values must be the same.

This particular case is on the very low end of the range where that would be called for, since the values only change once, but they're tested three times and there are 5 separate values compared, so I would probably make the change, but I'll leave it up to your judgement as to whether it's actually useful.  This approach is most useful when the tests and outcomes keep changing, and this particular case might be stable already.